File: Service/ContactService.vb
Imports BO
Imports Facade
Imports DAL

Public Class ContactService
    Implements IContactService

    Function GetContacts() As GetResponse(Of ContactBO) Implements IContactService.GetContacts
        Dim response As New GetResponse(Of ContactBO)

        Dim uow As New UnitOfWork(False)
        Dim dao = uow.GetCompanyContactsDAO()

        Dim entities = dao.GetNoTracking()
        For Each _entity In entities
            Dim bo As New ContactBO()
            Dim err = ContactTranslator.TranslateEntityToBO(_entity, bo)
            If err = ErrorCode.Success Then
                response.AddValue(bo)
            Else
                response.AddError(err.ToString())
            End If
        Next
        Return response
    End Function
    Function GetContactById(id As Integer) As GetResponse(Of ContactBO) Implements IContactService.GetContactById
        Dim response As New GetResponse(Of ContactBO)

        Dim uow As New UnitOfWork(False)
        Dim dao = uow.GetCompanyContactsDAO()
        Dim _entity = dao.GetById(id)
        Dim bo As New ContactBO()
        Dim err = ContactTranslator.TranslateEntityToBO(_entity, bo)
        If err = ErrorCode.Success Then
            response.AddValue(bo)
        Else
            response.AddError(err.ToString())
        End If

        Return response
    End Function
    Function GetContactsByIds(IdList As List(Of Integer)) As GetResponse(Of ContactBO) Implements IContactService.GetContactsByIds
        Dim response As New GetResponse(Of ContactBO)

        Dim uow As New UnitOfWork(False)
        Dim dao = uow.GetCompanyContactsDAO()
        For Each id In IdList
            Dim _entity = dao.GetById(id)
            Dim bo As New ContactBO()
            Dim err = ContactTranslator.TranslateEntityToBO(_entity, bo)
            If err = ErrorCode.Success Then
                response.AddValue(bo)
            Else
                response.AddError(err.ToString())
            End If
        Next
        Return response
    End Function
    Function GetContactsForSearchList(searchterm As String) As GetResponse(Of SelectBO) Implements IContactService.GetContactsForSearchList
        Dim response As New GetResponse(Of SelectBO)
        Dim uow As New UnitOfWork(False)
        Dim dao = uow.GetCompanyContactsDAO
        Dim entitys = dao.GetNoTracking().Where(ContactQuery.GetNameQuery(searchterm)).OrderBy(Function(m) m.ContactNaam).Select(Function(m) New SelectBO With {.id = m.CompanyID, .text = m.ContactNaam & " " & m.ContactVoornaam & " - " & m.CompanyInfo.BedrijfsNaam, .extra = "Contact"})
        response.Values = entitys.ToList()
        'For Each _entity In entitys
        '    response.AddValue(_entity.GetIdNameForSearch())
        'Next
        Return response
    End Function

    Function InsertUpdate(bo As ContactBO) As Response Implements IContactService.InsertUpdate
        Dim response As New Response()
        If (String.IsNullOrWhiteSpace(bo.Name)) Then
            response.AddError("name is mandatory")
        End If
        If (Not response.Success) Then Return response

        Dim uow As New UnitOfWork()
        Dim dao = uow.GetCompanyContactsDAO()
        Dim _entity As CompanyContacts = Nothing

        If (bo.Id = 0) Then
            _entity = dao.GetNew()
        Else
            _entity = dao.GetById(bo.Id)
        End If
        If (_entity IsNot Nothing) Then
            Dim err = ContactTranslator.TranslateBOToEntity(_entity, bo)
            If (err <> ErrorCode.Success) Then
                response.AddError(err.ToString())
            End If
        Else
            response.AddError("department not found")
        End If

        response.AddError(uow.SaveChanges())
        Return response
    End Function

    Function Delete(ids As List(Of Integer)) As Response Implements IContactService.Delete
        Dim response As New Response()
        Dim uow As New UnitOfWork()

        For Each id In ids
            uow.GetCompanyContactsDAO().DeleteObject(id)
        Next
        response.Messages.AddRange(uow.SaveChanges())

        Return response
    End Function
    Function Delete(bos As List(Of ContactBO)) As Response Implements IContactService.Delete
        Return Delete(bos.Select(Function(s) s.Id).ToList())
    End Function

End Class

File: ServiceCore/ContactService.cs
using BOCore;
using FacadeCore;
using DALCore;
using DALCore.Models;
using DALCore.Query;
using ServiceCore.Translators;
using System.Linq;

namespace ServiceCore
{
    public class ContactService : IContactService
    {
        private readonly UnitOfWorkCore _uow;

        public ContactService(UnitOfWorkCore uow)
        {
            _uow = uow;
        }

        public GetResponse<ContactBO> GetContacts()
        {
            var response = new GetResponse<ContactBO>();

            var entities = _uow.CompanyContacts.GetNoTracking();
            foreach (var e in entities)
            {
                var bo = new ContactBO();
                var err = ContactTranslator.TranslateEntityToBO(e, bo);
                if (err == ErrorCode.Success) response.AddValue(bo);
                else response.AddError(err.ToString());
            }
            return response;
        }

        public GetResponse<ContactBO> GetContactById(int id)
        {
            var response = new GetResponse<ContactBO>();

            var entity = _uow.CompanyContacts.GetById(id);
            var bo = new ContactBO();
            var err = ContactTranslator.TranslateEntityToBO(entity, bo);
            if (err == ErrorCode.Success) response.AddValue(bo);
            else response.AddError(err.ToString());

            return response;
        }

        public GetResponse<ContactBO> GetContactsByIds(List<int> idList)
        {
            var response = new GetResponse<ContactBO>();

            foreach (var id in idList)
            {
                var entity = _uow.CompanyContacts.GetById(id);
                var bo = new ContactBO();
                var err = ContactTranslator.TranslateEntityToBO(entity, bo);
                if (err == ErrorCode.Success) response.AddValue(bo);
                else response.AddError(err.ToString());
            }
            return response;
        }

        public GetResponse<SelectBO> GetContactsForSearchList(string searchterm)
        {
            var response = new GetResponse<SelectBO>();

            // Let op: ik ga ervan uit dat het contact een Id-veld heeft met de sleutel.
            // In je originele code stond 'CompanyId'; dat lijkt onjuist voor contacts.
            var query = _uow.CompanyContacts.GetNoTracking()
                .Where(ContactQuery.GetNameQuery(searchterm))
                .OrderBy(m => m.ContactNaam)
                .Select(m => new SelectBO
                {
                    id = m.CompanyId, // was m.CompanyId
                    text = m.ContactNaam + " " + m.ContactVoornaam + " - " + m.Company.BedrijfsNaam,
                    extra = "Contact"
                });

            response.Values = query.ToList();
            return response;
        }

        public Response InsertUpdate(ContactBO bo)
        {
            var response = new Response();
            if (string.IsNullOrWhiteSpace(bo.Name))
            {
                response.AddError("name is mandatory");
                return response;
            }

            CompanyContacts entity = bo.Id == 0
                ? _uow.CompanyContacts.GetNew()
                : _uow.CompanyContacts.GetById(bo.Id);

            if (entity == null)
            {
                response.AddError("contact not found");
                return response;
            }

            var err = ContactTranslator.TranslateBOToEntity(entity, bo);
            if (err != ErrorCode.Success)
            {
                response.AddError(err.ToString());
                return response;
            }

            var result = _uow.SaveChanges();
            response.AddSaveChangesResult(result, "Contact aangepast of toegevoegd", "Contact niet aangepast of toegevoegd");
            return response;
        }

        public Response Delete(List<int> ids)
        {
            var response = new Response();

            foreach (var id in ids)
                _uow.CompanyContacts.DeleteObject(id);

            var result = _uow.SaveChanges();
            response.AddSaveChangesResult(result, "Record(s) verwijderd", "Geen records verwijderd");
            return response;
        }

        public Response Delete(List<ContactBO> bos)
            => Delete(bos.Select(s => s.Id).ToList());
    }
}
